### PR TITLE
[#12517] Ensure that all necessary contact fields are synced

### LINKED ISSUE #12517: Syncing contacts
As a user 
I want to make sure contacts are fully synced
So that I can share contacts between my devices

### Description

Currently only added contacts are synced across devices.
We want to make sure any information about the contact is actually synced, which includes:

- ENS name
- Nickname
- Whether they are blocked
- Whether they have been removed

This will require changes only in status-go I believe.

### PR DESCRIPTION
**Pull request by rasom**
fix #12517

status: ready

### CODE REVIEW

**Comment by churik** [comment]
@rasom 
### ISSUE 1: Error when syncing multiaccounts with some data (couldn't find steps to reproduce)
Performed a test with syncing multiaccount with such data
1) messages in timeline
2) blocked, removed, added contacts with nicknames and ENS
3) group chats 
4) public chat
After device pairing and initial syncing got an error:
IMAGE 2021-09-09 16:40:24


Similar (but not the same) error I've got after blocking contact on paired device:
1) 2 devices are paired 
2) add a contact on device 1
3) block a contact on device 1
4) error appears on device 2
IMAGE 2021-09-09 16:52:46
That's all logs that I can gather when clicking `Report issue` from this state (seems useless):
[Status.log](https://github.com/status-im/status-react/files/7137583/Status.log)

### ISSUE 2: blocked contacts and contacts without ENS/or username added before pairing are not synced after initial sync 

**Steps:**
1) Device A: create multiaccount, 
  - add contact (without nickname)
  - block any user from public chat
  - add contact with ENS
  - add contact with nickname
2) Device B: restore same multiaccount
3) Advertise Device B to Device A
4) Sync Device A and Device B

**Expected result:**
can see all data from device A appeared after initial sync
**Actual result:**
no blocked user; 
no user without nickname
**OS:** IOS, Android
Logs from device A: [status_logs.zip](https://github.com/status-im/status-react/files/7137735/status_logs.zip)
Logs from device B: [Status-debug-logs.zip](https://github.com/status-im/status-react/files/7137738/Status-debug-logs.zip)

### ISSUE 3: Non-contact nicknames are not syncing
Maybe apart from PR, but desirable behavior as no need to add a user to contacts to set nickname for him.

**Steps:**
1) Device A: create multiaccount, set nickname to some user without adding him to contact list
2) Device B: restore same multiaccount
3) Sync Device A and Device B

**Expected result:**
can see nickname on Device B
**Actual result:**
no nickname. it appears only after adding user  a contact
**OS:** IOS, Android

**Comment by rasom** [comment]
ISSUE 1 fixed
ISSUE 3 not an issue, contacts which are not added are not synced atm
ISSUE 2 on it

**Comment by rasom** [comment]
@churik issue 2 should be fixed now
checking what can be done with issue 3

**Comment by churik** [comment]
### ISSUE 4: blocked users that are not added to contacts are not synced
**Steps:**
1) restore multiaccount from seed phrase on Device A and on device B
2) pair device
3) block some user from public chat on DeviceA


**Expected result:**
will be in blocked users on device B
Moreover, removing user from contacts (if it was added on device A) will remove it from blocked users on device B

**Actual result:**
not added to blocked users on  device B

**OS:** IOS, Android

**Comment by churik** [comment]
### ISSUE 5: contacts added after pairing are not shown on paired device and fetching history for public chats after adding contact becomes significantly slower on multiaccounts with some data

It is not reproducible for multiaccounts without data, checked on multiaccount with 5 contacts, several group/public chats and communities. 
Seems that it is related to issue 6, as after adding contact with ENS I started to experience issues with loading chat history

**Steps:**
1) restore multiaccount from seed phrase on Device A and on device B
2) pair devices
3) on device A add some user from public chats to contact
4) join any public chat

**Expected result:**
1. new contact will appear on device B
2. chat history will be fetched quickly (same as for new multiaccounts)

**Actual result:**
1. no contact on paired device
2. endless `loading` in public chat
[Status-debug-logs.zip](https://github.com/status-im/status-react/files/7144166/Status-debug-logs.zip)
IMAGE 2021-09-10 15:39:59

### ISSUE 6: contacts with ens names, added after sync are not synced and contacts without ENS are shown with @ on paired devices
photo_2021-09-10 16 41 33 jpeg 2021-09-10 16-42-09

**Comment by churik** [comment]
All previous issues are fixed, thank you @rasom !
Found new one:
### ISSUE 7: unread indicator is shown for own messages as new on paired device after syncing blocked contact

**Steps:**
1) restore multiaccount from seed phrase on device A and device B
2) pair devices 
3) device A: join public chat with history
4) device B: open another tab (i.e. wallet)
5) device A: start 1-1 chat with contact
6) device A: block any user from public chat
7) device A: send message to 1-1 chat 
4) device B: pay attention to unread indicator

**Expected result:**
no unread messages

**Actual result:**
sent messages are shown as unread

**OS:** IOS, Android
Monosnap 2021-09-14 17-09-05

**Comment by rasom** [comment]
ISSUE 7 is fixed

**Comment by churik** [comment]
### ISSUE 8: Mute and Unmute are messed up (IOS)

photo_2021-09-15 11 54 47 jpeg 2021-09-15 11-55-12

File: src/status_im/transport/message/core.cljs [hunk 1]
```diff
@@ -43,11 +43,12 @@
         ^js pin-messages               (.-pinMessages response-js)
         ^js removed-messages           (.-removedMessages response-js)
         sync-handler                   (when-not process-async process-response)]
-
     (cond
 
       (seq chats)
       (do
+        (when (seq contacts)
```
Comment (flexsurfer) [inline src/status_im/transport/message/core.cljs:50]: could you please add comment here

**Review by shivekkhurana — APPROVED** [review]

**Comment by qoqobolo** [comment]
### ISSUE 9: App crashes on the second paired device if mute some user on the primary device 

**Steps:**
1. Recover the same multiaccount on both devices (one of them is ios)
2. Sync devices
3. On the ios device: join a public chat, open some user's profile, and mute the user
4. Check the second device, the app crashes after a while

**Comment by rasom** [comment]
@qoqobolo fixed

**Comment by qoqobolo** [comment]
### ISSUE 10: Crash on the secondary device if block some user on the primary paired device (any OS)

**Steps:**

1. Recover the same multiaccount on both devices
2. Sync devices
3. Device1: join a public chat, open some user's profile, and block the user
4. Check Device2, the app crashes after a while

**Comment by rasom** [comment]
@qoqobolo fixed

File: src/status_im/transport/message/core.cljs (hunk 1 above)
Comment (rasom) [reply src/status_im/transport/message/core.cljs:50]: @flexsurfer sorry for delay, comment is added

**Pull request merged**